File: src-server/src/logbook/calls.js
import '/logbook/components.js';

import * as util from '/utils.js';

import * as cmd from '/commands.js';

import TextboxBtn from '/components/textbox-btn.js';

 

const ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';

 

export function init(app, html, module) {

  cmd.init(app, html, module);

}

 

// reset must be synchronous

export function reset(date = null) {

  cmd.reset();

 

  const dt = new Date().valueOf();

  if (date === null)

    date = util.getDate(dt);

 

  cmd.setState('dt', dt);

  cmd.setState('date', date);

  cmd.setState('columns', 0);

 

  resetQueue();

  resetTechs();

  resetLogbook();

  asyncReset();

}

 

async function asyncReset(date = cmd.useState('date')) {

  await cmd.asyncRegisterFile('logbook/' + date.toString(), resetLogbook);

  await cmd.asyncRegisterFile('study_queue', resetQueue);

  await cmd.asyncRegisterFile('technicians', resetTechs);

  cmd.registerAction({}, () => cmd.sendDB({}, 'logbook__active'));

  if (date === util.getDate(cmd.useState('dt'))) {

    cmd.registerAction({}, maskProgress);

  }

}

 

function resetLogbook() {

  // cmd.useState('date') is valueOf and not string of date

  cmd.assign('.ltm-index-head .dates-menu', {}, { date: cmd.useState('date') });

  document.querySelector('[data-name="ltm-index-textbox-date"]').dataset.data = new Date(cmd.useState('date')).toLocaleDateString();

  cmd.assign('.ltm-index-head .dates-head-lbl', { textContent: new Date(cmd.useState('date')).toLocaleDateString() }, { date: cmd.useState('date') });

  cmd.assign('.ltm-head-wrap', { textContent: '' });

  cmd.assign('.ltm-body-wrap', { textContent: '' });

  cmd.assign('.ltm-log-wrap', { textContent: '' });

  cmd.assign('.ltm-foot-wrap', { textContent: '' });

  for (let i = 0; i < 12; i++) {

    cmd.assign(`.ltm-timeslot:nth-child(${i + 1})`, { textContent: '' }, { timeslot: util.getTimeslot(cmd.useState('date'), i) });

  }

  cmd.assign('.ltm-summary', { textContent: '' });

  cmd.assign('.ltm-machines', { textContent: '' });

}

 

function resetQueue() {

  cmd.assign('.ltm-queue', { textContent: '' });

}

 

function resetTechs() {

  cmd.assign('.ltm-techs-menu', { textContent: '' });

}

 

// ----------------------------------- ACTIONS

 

function maskProgress(_) {

  const now = new Date().valueOf();

  for (const e of document.querySelectorAll(`.ltm-progress[data-timeslot="${util.getTimeslot(cmd.useState('date'), util.getSlot(now))}"] > .ltm-progress-mask`)) {

    e.style.setProperty('--inset-inline-start', util.getProgress(now));

  }

}

 

// ----------------------------------- DB CALLS

 

export function patient({ study, arrival }, mrn, first, last) {

  // if event from logbook else event from study_queue

  if (study) {

    cmd.assign(`.ltm-head[data-study="${study}"] > .ltm-name`, { textContent: `${first} ${last}` });

    cmd.assign(`.ltm-tbl-name[data-study="${study}"]`, { textContent: `${first} ${last}` });

  } else {

    cmd.assign(`.ltm-tbl-name[data-arrival="${arrival}"]`, { textContent: `${first} ${last}` });

  }

}

 

export function technician({ study, timeslot }, tech, first, last) {

  cmd.assign(`.ltm-timeslot[data-timeslot="${timeslot}"] > [data-study="${study}"]`, { textContent: `${first} ${last}` });

}

 

export function logbook__active(_, study, mrn, timer, status) {

  cmd.sendDB({ study }, 'patient', mrn);

  cmd.assign(`.ltm-head[data-study="${study}"] > .ltm-timer`, { textContent: util.formatTimer(timer) });

  cmd.assign(`.ltm-tbl-timer[data-study="${study}"]`, { textContent: util.formatTimer(timer) });

  assignStatus(study, status);

}

 

export function logbook(_, study, study_end, mrn, timer, status) {

  logbook__active(_, study, mrn, timer, status);

  // only set complete status when post_logbook_status event is not present in date's events

  if (study_end !== null && util.getDate(study_end) !== cmd.useState('date')) {

    assignCompleteStatus(study_end, study, status);

  }

}

 

// ----------------------------------- FILE CALLS

 

export function post_technician(tech, first, last) {

  cmd.append('.ltm-techs-menu', 'ltm-techs-item', { className: 'float-menu-item ltm-techs-item', textContent: `${first} ${last}` }, { tech });

}

 

export function post_study_queue(arrival, mrn) {

  cmd.append('.ltm-queue', 'div', { className: 'ltm-tbl-item', textContent: mrn }, { arrival, mrn });

  cmd.append('.ltm-queue', 'div', { className: 'ltm-tbl-name ltm-tbl-item' }, { arrival, mrn });

  cmd.append('.ltm-queue', 'div', { className: 'ltm-tbl-arrival ltm-tbl-item', textContent: util.formatDatetime(arrival) }, { arrival, mrn });

  cmd.append('.ltm-queue', 'ltm-start-btn', { className: 'btn ltm-start-btn ltm-tbl-item', textContent: 'Start' }, { arrival, mrn });

  cmd.sendDB({ arrival }, 'patient', mrn);

}

 

export function post_logbook(study, study_end, mrn, timer, status, ...events) {

  cmd.setState(study, true);

  cmd.setState('columns', cmd.useState('columns') + 1);

  document.querySelector('.ltm-date').style.setProperty('--columns', cmd.useState('columns'));

  cmd.append('.ltm-head-wrap', 'div', { className: 'ltm-head', innerHTML: `<div class="ltm-timer">00:00:00</div><div class="ltm-name"></div><div class="ltm-mrn">${mrn}</div><ltm-timer-btn class="btn ltm-timer-btn"></ltm-timer-btn>` }, { study })

  cmd.append('.ltm-foot-wrap', 'div', { className: 'ltm-foot', innerHTML: '<div class="float"><div class="float-head"><ltm-events-btn class="btn blue float-head-lbl ltm-events-btn">Events</ltm-events-btn><float-head-btn class="btn blue float-head-btn"><svg class="float-head-btn-img" xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path fill="currentColor" d="M3 19h18a1.002 1.002 0 0 0 .823-1.569l-9-13c-.373-.539-1.271-.539-1.645 0l-9 13A.999.999 0 0 0 3 19z" /></svg></float-head-btn></div><div class="float-menu-wrap"><div class="float-menu up ltm-events-menu"><ltm-events-item data-item="0" class="float-menu-item ltm-events-item">A</ltm-events-item><ltm-events-item data-item="1" class="float-menu-item ltm-events-item">B</ltm-events-item><ltm-events-item data-item="2" class="float-menu-item ltm-events-item">C</ltm-events-item><ltm-events-item data-item="3" class="float-menu-item ltm-events-item">D</ltm-events-item><ltm-events-item data-item="4" class="float-menu-item ltm-events-item">E</ltm-events-item><ltm-events-item data-item="5" class="float-menu-item ltm-events-item">F</ltm-events-item><ltm-events-item data-item="6" class="float-menu-item ltm-events-item">G</ltm-events-item><ltm-events-item data-item="7" class="float-menu-item ltm-events-item">H</ltm-events-item><ltm-events-item data-item="8" class="float-menu-item ltm-events-item">I</ltm-events-item><ltm-events-item data-item="9" class="float-menu-item ltm-events-item">J</ltm-events-item><ltm-events-item data-item="10" class="float-menu-item ltm-events-item">K</ltm-events-item><ltm-events-item data-item="11" class="float-menu-item ltm-events-item">L</ltm-events-item><ltm-events-item data-item="12" class="float-menu-item ltm-events-item">M</ltm-events-item><ltm-events-item data-item="13" class="float-menu-item ltm-events-item">N</ltm-events-item><ltm-events-item data-item="14" class="float-menu-item ltm-events-item">O</ltm-events-item><ltm-events-item data-item="15" class="float-menu-item ltm-events-item">P</ltm-events-item><ltm-events-item data-item="16" class="float-menu-item ltm-events-item">Q</ltm-events-item><ltm-events-item data-item="17" class="float-menu-item ltm-events-item">R</ltm-events-item><ltm-events-item data-item="18" class="float-menu-item ltm-events-item">S</ltm-events-item><ltm-events-item data-item="19" class="float-menu-item ltm-events-item">T</ltm-events-item><ltm-events-item data-item="20" class="float-menu-item ltm-events-item">U</ltm-events-item><ltm-events-item data-item="21" class="float-menu-item ltm-events-item">V</ltm-events-item><ltm-events-item data-item="22" class="float-menu-item ltm-events-item">W</ltm-events-item><ltm-events-item data-item="23" class="float-menu-item ltm-events-item">X</ltm-events-item><ltm-events-item data-item="24" class="float-menu-item ltm-events-item">Y</ltm-events-item><ltm-events-item data-item="25" class="float-menu-item ltm-events-item">Z</ltm-events-item></div></div></div><ltm-complete-btn class="btn green disabled ltm-complete-btn">Complete</ltm-complete-btn>' }, { study });

  cmd.append('.ltm-log-wrap', 'div', { className: 'ltm-log' }, { study });

  cmd.append('.ltm-body-wrap', 'div', { className: 'ltm-body', innerHTML: `<div class="ltm-sign-btn-wrap"><ltm-sign-btn class="disabled btn ltm-sign-btn">Sign</ltm-sign-btn></div><div data-timeslot="${util.getTimeslot(cmd.useState('date'), 0)}" class="ltm-progress"><ltm-progress-mask class="ltm-progress-mask"></ltm-progress-mask></div><div data-timeslot="${util.getTimeslot(cmd.useState('date'), 1)}" class="ltm-progress"><ltm-progress-mask class="ltm-progress-mask"></ltm-progress-mask></div><div data-timeslot="${util.getTimeslot(cmd.useState('date'), 2)}" class="ltm-progress"><ltm-progress-mask class="ltm-progress-mask"></ltm-progress-mask></div><div data-timeslot="${util.getTimeslot(cmd.useState('date'), 3)}" class="ltm-progress"><ltm-progress-mask class="ltm-progress-mask"></ltm-progress-mask></div><div data-timeslot="${util.getTimeslot(cmd.useState('date'), 4)}" class="ltm-progress"><ltm-progress-mask class="ltm-progress-mask"></ltm-progress-mask></div><div data-timeslot="${util.getTimeslot(cmd.useState('date'), 5)}" class="ltm-progress"><ltm-progress-mask class="ltm-progress-mask"></ltm-progress-mask></div><div data-timeslot="${util.getTimeslot(cmd.useState('date'), 6)}" class="ltm-progress"><ltm-progress-mask class="ltm-progress-mask"></ltm-progress-mask></div><div data-timeslot="${util.getTimeslot(cmd.useState('date'), 7)}" class="ltm-progress"><ltm-progress-mask class="ltm-progress-mask"></ltm-progress-mask></div><div data-timeslot="${util.getTimeslot(cmd.useState('date'), 8)}" class="ltm-progress"><ltm-progress-mask class="ltm-progress-mask"></ltm-progress-mask></div><div data-timeslot="${util.getTimeslot(cmd.useState('date'), 9)}" class="ltm-progress"><ltm-progress-mask class="ltm-progress-mask"></ltm-progress-mask></div><div data-timeslot="${util.getTimeslot(cmd.useState('date'), 10)}" class="ltm-progress"><ltm-progress-mask class="ltm-progress-mask"></ltm-progress-mask></div><div data-timeslot="${util.getTimeslot(cmd.useState('date'), 11)}" class="ltm-progress"><ltm-progress-mask class="ltm-progress-mask"></ltm-progress-mask></div>` }, { study });

 

  cmd.append('.ltm-summary', 'div', { className: 'ltm-tbl-item', textContent: mrn }, { study });

  cmd.append('.ltm-summary', 'div', { className: 'ltm-tbl-name ltm-tbl-item' }, { study });

  cmd.append('.ltm-summary', 'div', { className: 'ltm-tbl-timer ltm-tbl-item' }, { study });

  cmd.append('.ltm-summary', 'ltm-show-btn', { className: 'btn ltm-show-btn ltm-tbl-item active' }, { study });

  // update mrn, timer, and status for previous dates of complete studies

  let timeslot;

  if (study >= cmd.useState('date')) {

    const slot = util.getSlot(study);

    timeslot = util.getTimeslot(cmd.useState('date'), slot);

    document.querySelector(`.ltm-body[data-study="${study}"] .ltm-sign-btn`).dataset.slot = slot;

  } else {

    timeslot = cmd.useState('date');

  }

  cmd.append(`.ltm-body[data-study="${study}"] > .ltm-progress[data-timeslot="${timeslot}"]`, 'ltm-progress-bar', { className: `ltm-progress-bar` }, { dt: timeslot });

  cmd.sendDB({}, 'logbook', study);

 

  for (const [i, description] of Object.entries(events)) {

    if (description === null) break;

    post_logbook__event(description, study, i);

  }

}

 

export function post_logbook_status(dt, study, status) {

  appendStatus(dt, study, status, util.getTimeslot(cmd.useState('date'), util.getSlot(dt)));

  assignStatus(study, status);

  if (status.includes('complete')) {

    assignCompleteStatus(dt, study, status);

  }

}

 

export function post_logbook_signature(dt, study, tech, timeslot, eeg, video, maintenance) {

  // enable complete btn when all timeslots signed

  if (timeslot === util.getTimeslot(dt)) {

    document.querySelector(`.ltm-foot[data-study="${study}"] > .ltm-complete-btn`).classList.remove('disabled');

  }

  document.querySelector(`.ltm-body[data-study="${study}"] .ltm-sign-btn`).dataset.slot = util.getSlot(timeslot) + 1;

  cmd.sendDB({ study, timeslot }, 'technician', tech);

}

 

export function post_logbook_event(dt, study, i) {

  cmd.append(`.ltm-log[data-study="${study}"]`, 'div', { className: 'ltm-log-item', textContent: util.formatTime(dt) }, { event: ALPHABET.at(i) });

}

 

export function post_logbook__event(description, study, i) {

  cmd.assign(`.ltm-foot[data-study="${study}"] [data-item="${i}"]`, {}, { description });

}

 

export function timeslot(timeslot, study) {

  cmd.append(`.ltm-timeslot[data-timeslot="${timeslot}"]`, 'div', { className: 'ltm-sign' }, { study });

  document.querySelector(`.ltm-body[data-study="${study}"] .ltm-sign-btn`).dataset.timeslotSlot = util.getSlot(timeslot);

  document.querySelector(`.ltm-foot[data-study="${study}"] > .ltm-complete-btn`).classList.add('disabled');

}

 

// ----------------------------------- LOCAL

 

function appendStatus(dt, study, status, timeslot) {

  cmd.append(`.ltm-body[data-study="${study}"] > .ltm-progress[data-timeslot="${timeslot}"]`, 'ltm-progress-bar', { className: `ltm-progress-bar ${status}` }, { dt });

}

 

function assignStatus(study, status) {

  const [statusLtm, statusTimer] = status.split(' ');

  cmd.assign(`.ltm-head[data-study="${study}"] > .ltm-timer-btn`, {}, { statusLtm, statusTimer });

  cmd.assign(`.ltm-foot[data-study="${study}"] > .ltm-complete-btn`, {}, { statusLtm, statusTimer });

  if (statusLtm === 'ambulatory' || statusLtm === 'unmonitored') {

    disableButtons(study);

  }

}

 

function assignCompleteStatus(dt, study, status) {

  disableButtons(study);

  for (let i = util.getSlot(dt); i < 12; i++) {

    document.querySelector(`.ltm-body[data-study="${study}"] > .ltm-progress[data-timeslot="${util.getTimeslot(util.getDate(dt), i)}"] > .ltm-progress-mask`).classList.add('complete');

    appendStatus(dt, study, status, util.getTimeslot(util.getDate(dt), i));

  }

}

 

function disableButtons(study) {

  document.querySelector(`.ltm-body[data-study="${study}"] > .ltm-sign-btn-wrap`).classList.add('disabled');

  document.querySelector(`.ltm-head[data-study="${study}"] > .ltm-timer-btn`).classList.add('disabled');

  document.querySelector(`.ltm-foot[data-study="${study}"] > .ltm-complete-btn`).classList.add('disabled');

  document.querySelector(`.ltm-foot[data-study="${study}"] .ltm-events-btn`).classList.add('disabled');

  document.querySelector(`.ltm-foot[data-study="${study}"] .float-head-btn`).classList.add('disabled');

}

 

// -----------------------------------

 

class DatesMenuBodyItem extends HTMLElement {

    connectedCallback() {

        this.addEventListener("click", this.onclick);

    }

    disconnectedCallback() {

        this.removeEventListener("click", this.onclick);

    }

    onclick() {

        reset(+this.dataset.date);

    }

}

 

class IndexTextboxDate extends TextboxBtn {

 

    validate_input(text) {

        if (this.dataset.data === undefined) {

            return new Date().toLocaleDateString();

        }

 

        // day: 1-31, month: 1-12

        // date.day: 1-31, date.month: 0-11

        const oldSplit = this.dataset.data.split('/');

        const split = text.split('/');

        const [month, day, year] = split.map((field, i) => field.length ? field : oldSplit.at(i));

 

        if (split.length !== 3) {

            return this.dataset.data;

        }

        else if (isNaN(day) || isNaN(month) || isNaN(year)) {

            return this.dataset.data;

        }

        else if (+year <= 1970 || +year > 3000) {

            return this.dataset.data;

        }

        else if (+month < 1 || +month > 12) {

            return this.dataset.data;

        }

        else if (+day < 1 || +day > util.getMaxDay(+month - 1, +year)) {

            return this.dataset.data;

        }

        else {

            return `${util.stripDouble(month)}/${util.stripDouble(day)}/${year}`;

        }

    }

 

    async submit_input() {

        const [month, day, year] = this.dataset.data.split('/').map(field => +field);

        const date = new Date(year, month - 1, day);

 

        if (await cmd.asyncIsFile(`logbook\\${date.valueOf()}`)) {

            this.dataset.data = date.toLocaleDateString();

            reset(date.valueOf());

        }

        else {

            this.dataset.data = new Date(+this.dataset.date).toLocaleDateString();

        }

    }

}

 

customElements.define('ltm-index-textbox-date', IndexTextboxDate);

customElements.define("dates-menu-body-item", DatesMenuBodyItem);

File: src-server/logbook/calls.py
import utils
import os

 

ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'

MILLI = 1000




# --------------------------------- LOGBOOK ---------------------------------

 

def logbook(cursor, send_files, args, row):

    """

        study (int),

    """

    cursor.execute("SELECT * FROM logbook WHERE study = %s;", row)

    return cursor.fetchall()

 

def post_logbook(cursor, send_files, args, row):

    """

        queue (int), # not in logbook!

       

        study (int),

        study_end (int),

        mrn (int),

        timer (int),

        status (str),

        event_a (str), # all events are optional

        ...

        event_z (str),

    """

    date = utils.get_date_from(row[1] // MILLI)  * MILLI

    timeslot = utils.get_timeslot_from(row[1] // MILLI) * MILLI

    prev_date_last_timeslot = utils.get_prev_timeslot_from(date // MILLI) * MILLI

    filepath = f'logbook/{date}'

    row_values = row[1:]

    row_values.extend([None for _ in range(32 - len(row))])

    cursor.execute("SELECT MAX (timeslot) FROM logbook_signature;")

    sign_timeslot = cursor.fetchone()[0]

   

    cursor.execute("INSERT INTO logbook VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s);", row_values)

    cursor.execute("INSERT INTO logbook_status VALUES (%s, %s, %s);", (row[1], row[1], row[5]))

    cursor.execute("UPDATE study_queue SET logbook_study = %s WHERE queue = %s;", (row[1], row[0]))

    refresh_study_queue(cursor, send_files)

    send_files(

        'a',

        filepath,

        'post_logbook',

        [row_values]

    )

    send_files(

        'a',

        filepath,

        'post_logbook_status',

        [(row[1], row[1], row[5])]

    )

 

    # if no previous signatures

    #   if study date >= min unsigned date

    #       - post timeslot

    # if previous date is all signed

    #   - post timeslot

 

    if sign_timeslot is None:

        cursor.execute("SELECT MIN (study) FROM logbook;")

        min_study = cursor.fetchone()[0]

        min_study = min_study if min_study is not None else row[1]

        min_study_date = utils.get_date_from(min_study // MILLI) * MILLI

        # if date >= min unsigned date

        if date >= min_study_date:

            send_files(

                'a',

                filepath,

                'timeslot',

                [(timeslot, row[1])]

            )

    # if prev date is all signed

    elif prev_date_last_timeslot <= sign_timeslot and utils.is_timeslot_signed(cursor, prev_date_last_timeslot):

        send_files(

            'a',

            filepath,

            'timeslot',

            [(timeslot, row[1])]

        )

 

def logbook__active(cursor, send_files, args, row):

    """

        study_end (None),

    """

    cursor.execute("SELECT * FROM logbook WHERE study_end IS NULL ORDER BY study;")

    return [(row[0], row[2], row[3], row[4]) for row in cursor.fetchall()]

 

def post_logbook__event(cursor, send_files, args, row):

    """

        event (int),            zero-indexed int of letter of event

        event_{a, ...z} (str),  event description

        study (int),

    """

    if type(row[0]) is int:

        post = f'UPDATE logbook SET event_{  ALPHABET[row[0]] } = %s WHERE study = %s;'

        cursor.execute(post, row[1:])

        send_files('a', f'logbook/{utils.get_date() * MILLI}', 'post_logbook__event', [(*row[1:], row[0])])

 

def refresh_logbook(cursor, send_files) -> None:

 

    cursor.execute("SELECT MIN (study) FROM logbook;")

    min_study = cursor.fetchone()[0]

    if min_study is None:

        return open(f'data/logbook/{utils.get_date() * MILLI}', 'w').close()

 

    events = {}

    now_timeslot = utils.get_timeslot() * MILLI

    now_date = utils.get_date_from(now_timeslot // MILLI) * MILLI

    dates = utils.create_logbook_dates(min_study, now_date)

    events = utils.add_event_timeslot_keys(min_study, events, now_timeslot)

    max_timeslot = utils.get_max_timeslot(cursor, now_timeslot)

 

    cursor.execute("SELECT * FROM logbook ORDER BY study;")

    for row in cursor.fetchall():

 

        start_date = utils.get_date_from(row[0] // MILLI) * MILLI

        start_timeslot = utils.get_timeslot_from(row[0] // MILLI) * MILLI

        end_timeslot = now_timeslot if row[1] is None else utils.get_timeslot_from(row[1] // MILLI) * MILLI

        end_date = utils.get_date_from(end_timeslot // MILLI) * MILLI

        end_date = utils.get_next_date_from(end_date // MILLI) * MILLI

        end_date_timeslot = utils.get_prev_timeslot_from(end_date // MILLI) * MILLI

        end_date_timeslot = min(max_timeslot, end_date_timeslot)

 

        events = utils.add_event_dates(events, start_date, end_date, row)

        events = utils.add_event_statuses(cursor, events, start_timeslot, end_timeslot, row)

        events = utils.add_event_timeslots(events, start_timeslot, end_date_timeslot, row[0])

 

    for date in dates:

        filepath = f'logbook/{date}'

        next_date = utils.get_next_date_from(date // MILLI) * MILLI

 

        cursor.execute("SELECT * FROM logbook_signature WHERE timeslot BETWEEN %s AND %s ORDER BY timeslot;", (date, next_date))

        events = utils.map_dt_events(events, cursor.fetchall(), filepath, 'post_logbook_signature')

        cursor.execute("SELECT * FROM logbook_status WHERE dt BETWEEN %s AND %s ORDER BY dt;", (date, next_date))

        events = utils.map_dt_events(events, cursor.fetchall(), filepath, 'post_logbook_status')

        cursor.execute("SELECT * FROM logbook_event WHERE dt BETWEEN %s AND %s ORDER BY dt;", (date, next_date))

        events = utils.map_dt_events(events, cursor.fetchall(), filepath, 'post_logbook_event')

 

    for dt in sorted(events):

        for args in events[dt]:

            send_files(*args)

 

# --------------------------------- LOGBOOK EVENT ---------------------------------

 

def post_logbook_event(cursor, send_files, args, row):

    """

        dt (int),

        study (int),

        event (int),

    """

    cursor.execute("INSERT INTO logbook_event VALUES (%s, %s, %s);", row)

    send_files('a', f'logbook/{utils.get_date_from(row[0] // MILLI) * MILLI}', 'post_logbook_event', [row])

 

# --------------------------------- LOGBOOK SIGNATURE ---------------------------------

 

def post_logbook_signature(cursor, send_files, args, row):

    """

        dt (int) ms,

        study (int) ms,

        tech (int) ms,

        timeslot (int) ms,

        eeg_integrity_quality (bool),

        video_integrity_quality (bool),

        needs_maintenance (bool),

    """

    cursor.execute("INSERT INTO logbook_signature VALUES (%s, %s, %s, %s, %s, %s, %s);", row)

    timeslot = row[3]

    date = utils.get_date_from(timeslot // MILLI) * MILLI

    next_date = utils.get_next_date_from(date // MILLI) * MILLI

    date_end_timeslot = utils.get_prev_timeslot_from(next_date // MILLI) * MILLI

    send_files('a', f'logbook/{date}', 'post_logbook_signature', [row])

 

    cursor.execute("SELECT MAX (timeslot) FROM logbook_signature;")

    sign_timeslot = cursor.fetchone()[0]

    sign_timeslot = sign_timeslot if sign_timeslot is not None else timeslot

 

    # if timeslot == sign_timeslot

    # and timeslot is last timeslot of date

    # and timeslot is all signed

    #   - entire date gets filled when next_date isn't today

    #   - else date gets filled upto current_timeslot

 

    if timeslot == sign_timeslot and \
        timeslot == date_end_timeslot and \
        utils.is_timeslot_signed(cursor, timeslot):

 

        next_next_date = utils.get_next_date_from(next_date // MILLI) * MILLI

        next_date_last_timeslot = utils.get_prev_timeslot_from(next_next_date // MILLI) * MILLI

        max_timeslot = min(next_date_last_timeslot, utils.get_timeslot() * MILLI)

        send_files(

            'a',

            f'logbook/{next_date}',

            'timeslot',

            utils.collect_timeslots(cursor, next_date, max_timeslot)

        )

 

# --------------------------------- LOGBOOK STATUS ---------------------------------

 

def post_logbook_status(cursor, send_files, args, row):

    """

        dt (int),

        study (int),

        status (str)

    """

    cursor.execute("UPDATE logbook SET status = %s WHERE study = %s;", (row[2], row[1]))

    cursor.execute("INSERT INTO logbook_status VALUES (%s, %s, %s);", row)

    send_files(

        'a',

        f'logbook/{utils.get_date_from(row[0] // 1000)  * MILLI}',

        'post_logbook_status',

        [row]

    )

    if 'complete' == row[2].split(' ')[1]:

        cursor.execute("UPDATE logbook SET study_end = %s WHERE study = %s;", (row[0], row[1]))

def patient(cursor, send_files, args, row):

    """

    row:

        mrn (int),

    """

    cursor.execute("SELECT * FROM patient WHERE mrn = %s;", row)

    return cursor.fetchall()

 

def post_patient(cursor, send_files, args, row):

    """

    row:

        mrn (int),

        first_name (str),

        last_name (str),

    """

    cursor.execute("INSERT INTO patient VALUES (%s, %s, %s);", row)

 

# --------------------------------- TECHNICIAN ---------------------------------

 

def technician(cursor, send_files, args, row):

    """

    row:

        tech (int),

    """

    cursor.execute("SELECT * FROM technician WHERE tech = %s;", row)

    return cursor.fetchall()

 

def post_technician(cursor, send_files, args, row):

    """

    row:

        tech (int),

        first_name (str),

        last_name (str)

    """

    cursor.execute("INSERT INTO technician VALUES (%s, %s, %s, NULL) RETURNING *;", row)

    send_files('a', 'technicians', 'post_technician', cursor.fetchall())

 

def refresh_technician(cursor, send_files):

    cursor.execute("SELECT * FROM technician WHERE del IS NULL ORDER BY tech;")

    send_files('w', 'technicians', 'post_technician', cursor.fetchall())

 

def post_technician__del(cursor, send_files, args, row):

    """

    row:

        del (int),

        tech (int),

    """

    cursor.execute("UPDATE technician SET del = %s WHERE tech = %s;", row)

    refresh_technician(cursor, send_files)

 

# --------------------------------- STUDY QUEUE ---------------------------------

 

def post_study_queue(cursor, send_files, args, row):

    """

    row:

        queue (int) datetime arrived,

        mrn (int),

    """

    cursor.execute("INSERT INTO study_queue VALUES (%s, %s, NULL) RETURNING *;", row)

    send_files('a', 'study_queue', 'post_study_queue', cursor.fetchall())

 

def refresh_study_queue(cursor, send_files):

    cursor.execute("SELECT * FROM study_queue WHERE logbook_study IS NULL ORDER BY queue;")

    send_files('w', 'study_queue', 'post_study_queue', cursor.fetchall())